core/src/state/binding.rs

```diff
@@ -1,141 +1,264 @@
+        cx.views.insert(id, Box::new(binding));
+
+        cx.with_current(id, |cx| {
+            // Call the body of the binding
+            if let Some(mut view_handler) = cx.views.remove(&id) {
+                view_handler.body(cx);
+                cx.views.insert(id, view_handler);
+            }
+        });
+
+        let _: Handle<Self> = Handle { entity: id, p: Default::default(), cx }.ignore();
+    }
+}
+
+impl<F: 'static + Fn(&mut Context, L), L: Bindable> View for Binding<F, L> {
+    fn body(&mut self, cx: &mut Context) {
+        cx.remove_children(cx.current());
+        (self.content)(cx, self.bindable.clone());
+    }
+}
+
+fn alloc_observer<L: Bindable>(cx: &mut Context, bindable: L) -> Entity {
+    let current = cx.current();
+    let id = cx.entity_manager.create();
+    cx.tree().add(id, current).expect("Failed to add to tree");
+    cx.cache().add(id).expect("Failed to add to cache");
+    cx.style().add(id);
+
+    let new_ancestors = id.parent_iter(cx.tree()).collect::<Vec<_>>();
+
+    // for each model the bindable requests...
+    for store_req in bindable.requests() {
+        // walk the ancestry chain child-to-parent...
+        for entity in new_ancestors.iter() {
+            // key the data map on the anchor entity...
+            if let Some(model_data_store) = cx.data.get_mut(*entity) {
+                // and then the model type...
+                let model_data =
+                    if let Some(model_data) = model_data_store.data.get(&store_req.model_id) {
+                        ModelOrView::Model(model_data.as_ref())
+                    } else if let Some(view_data) = cx.views.get(entity).and_then(|view| {
+                        ((view.as_any_ref()).type_id() == store_req.model_id).then(|| view)
+                    }) {
+                        ModelOrView::View(view_data.as_ref())
+                    } else {
+                        continue;
+                    };
+
+                // and finally the store type...
+                // (but only if it's dedupable)
+                let store = if store_req.can_deduplicate {
+                    model_data_store.lenses_dedup.entry(store_req.store_id).or_insert_with(|| {
+                        bindable.make_store(model_data).expect(
+                            "Bug: bindable failed to produce a store. Probably an internal error.",
+                        )
+                    })
+                } else {
+                    let idx = model_data_store.lenses_dup.len();
+                    model_data_store.lenses_dup.push(bindable.make_store(model_data).expect(
+                        "Bug: bindable failed to produce a store. Probably an internal error.",
+                    ));
+                    model_data_store.lenses_dup.get_mut(idx).unwrap()
+                };
+
+                // ...so that we can register the observer with the appropriate store.
+                bindable.add_to_store(store.as_mut(), id);
```

Comment:
Why is this a method on `bindable` rather than on `store`? Would it not make more sense to have `store.add_observer(id)`?

Reply:
Ah I see! But is the observer not always an entity?

Reply:
The observer will always be an entity, but the observer's metadata (e.g. the list index being bound to) can be of arbitrary type.